From 87271eba50dec8151ac12b524bc2f88c10dd6079 Mon Sep 17 00:00:00 2001
From: John Peck <peck@bay2sierra.com>
Date: Mon, 12 Oct 2015 21:42:53 -0700
Subject: [PATCH] Added OPENSSL_NO_SHA512 to restrict use of SHA512 and SHA384

---
 ssl/t1_lib.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/ssl/t1_lib.c b/ssl/t1_lib.c
index f18f502..c5657f0 100644
--- a/ssl/t1_lib.c
+++ b/ssl/t1_lib.c
@@ -949,8 +949,10 @@ static int tls1_check_cert_param(SSL *s, X509 *x, int set_ee_md)
                 tlsext_sigalg_ecdsa(md)
 
 static const unsigned char tls12_sigalgs[] = {
+# ifndef OPENSSL_NO_SHA512
     tlsext_sigalg(TLSEXT_hash_sha512)
         tlsext_sigalg(TLSEXT_hash_sha384)
+# endif
         tlsext_sigalg(TLSEXT_hash_sha256)
         tlsext_sigalg(TLSEXT_hash_sha224)
         tlsext_sigalg(TLSEXT_hash_sha1)
-- 
1.9.1

